qiskit_experiments/framework/base_experiment.py

```diff
@@ -167,29 +144,144 @@ def _initialize_experiment_data(
+
+            Note that the analysis routine will not complete until the backend job
+            is executed, and this method will be called after the analysis routine
+            is completed though a handler of the experiment result will be immediately
+            returned to users (a future object). This method is automatically triggered
+            when the analysis is finished, and will be processed in background.
+
+            If this method updates some other (mutable) objects, you may need manage
+            synchronization of update of the object data. Otherwise you may want to
+            call :meth:`block_for_results` method of the ``experiment_data`` here
+            to freeze processing chain until the job result is returned.
+
+            By default, this method does nothing.
+
+        Args:
+            experiment_data: A future object of the experimental result.
+        """
+        pass
+
+    def run_analysis(
+        self, experiment_data: ExperimentData, job: BaseJob = None, **options
+    ) -> ExperimentData:
         """Run analysis and update ExperimentData with analysis result.
 
         Args:
-            experiment_data (ExperimentData): the experiment data to analyze.
-            options: additional analysis options. Any values set here will
-                     override the value from :meth:`analysis_options`
-                     for the current run.
+            experiment_data: The experiment data to analyze.
+            job: The future object of experiment result which is currently running on the backend.
+            options: Additional analysis options. Any values set here will
+                override the value from :meth:`analysis_options` for the current run.
 
         Returns:
             An experiment data object containing the analysis results and figures.
 
         Raises:
-            QiskitError: if experiment_data container is not valid for analysis.
+            QiskitError: Method is called with an empty experiment result.
         """
+        run_analysis = self.analysis() if self.__analysis_class__ else None
+
         # Get analysis options
         analysis_options = copy.copy(self.analysis_options)
         analysis_options.update_options(**options)
         analysis_options = analysis_options.__dict__
 
-        # Run analysis
-        analysis = self.analysis()
-        analysis.run(experiment_data, **analysis_options)
+        if not job and run_analysis is not None:
+            # Run analysis immediately
+            if not experiment_data.data():
+                raise QiskitError(
+                    "Experiment data seems to be empty and no running job is provided. "
+                    "At least one data entry is required to run analysis."
+                )
+            experiment_data = run_analysis.run(experiment_data, **analysis_options)
+        else:
+            # Run analysis when job is completed
+            experiment_data.add_data(
```

Comment:
This gives this function a confusing signature, if i call it with data and a job it is going to try and add that job to the data, but adding job data to experiment data should probably always be done by the `run` method, not the `run_analysis` method.

Reply:
This is fixed in https://github.com/Qiskit/qiskit-experiments/pull/380/commits/bf34b7793500e301f08a35ca315d6ab7af449ae1